### ISSUE TITLE
Ahead of time (AOT) compilation support 

### ISSUE THREAD

**Issue by itrethan** [AngularClass/angular2-webpack-starter#885]
**Note: for support questions, please use one of these channels:** [Chat: AngularClass.slack](http://angularclass.com/member-join/) or [Twitter: @AngularClass](https://twitter.com/AngularClass)
- **I'm submitting a ...**
  [ ] bug report
  [x ] feature request
  [ ] question about the decisions made in the repository
- **Do you want to request a _feature_ or report a _bug_?**
  Request a feature, AOT support in webpack
- **What is the current behavior?**
  JIT compile, performance is bad in prod.
- **What is the motivation / use case for changing the behavior?**
  Performance boost

**Comment by IAMtheIAM** [comment]
I agree, we should use AOT compilation now that it's available in Angular2!

**Comment by unsafecode** [comment]
+1

**Comment by qdouble** [comment]
there's still a lot of bugs and peculiarities in aot that would break most people's app...so this doesn't need to be added immediately

**Comment by IAMtheIAM** [comment]
At least we can get a writeup explaining how to do it individually, so we can work the kinks out ourselves.

**Comment by itrethan** [comment]
The details could be found here: http://blog.mgechev.com/2016/08/14/ahead-of-time-compilation-angular-offline-precompilation/

Just wondering how to do that with webpack, maybe another loader?

**Comment by Spittal** [comment]
+1

**Comment by mdkholy** [comment]
+1

**Comment by qdouble** [comment]
I was finally able to get AOT running fine with Webpack 2, you guys can check it here: https://github.com/qdouble/angular2webpack2-starter . There is a known issue with lazy routes when using AOT though, hopefully they patch it soon https://github.com/angular/angular/issues/11075

**Comment by IAMtheIAM** [comment]
@qdouble I downloaded your repo and got it to compile, excellent!. Only problem is, it doesn't work with any `require()` statements, like used in most webpack apps. It just gives error:

`Error: Error encountered resolving symbol values statically. Calling function 'require', function calls are not supported. Consider replacing the function or lambda with a reference to an exported function, `

**Comment by qdouble** [comment]
@IAMtheIAM that's an AOT limitation. You can't use require on your styles or templates with offline compile. It should work fine with JIT. For your html and styles, use templateUrl and stylesUrls... they will be bundled properly because the starter is using the angular2-template-loader.

**Comment by qdouble** [comment]
I've updated my starter project to fully support lazy loading in AOT mode. I have to wait until there are some more updates/bug fixes before making it a bit cleaner, but it's a viable solution for now: 
https://github.com/qdouble/angular2webpack2-starter/commit/b6bdf6e259c9d26ebb370ce9984f2122e46a0782

**Comment by Enlcxx** [comment]
+1

**Comment by kodeine** [comment]
@qdouble any updates on this ?

**Comment by qdouble** [comment]
@kodeine uh, my angular repo has grown and is working very well, but I'm not trying to say anything negative about this repo... my link is here: https://github.com/qdouble/angular2webpack2-starter address any issues there.

**Comment by breslavsky** [comment]
+1. AOT very important to increase speed for first start.
We have big app on angular2 and we have first init time about 2-3 sec. after 1 sec of loading.
4 seconds is not good for reactive apps. Thanks.

**Comment by vadimzham** [comment]
@qdouble I think, you should make more production-ready config for webpack. I mean splitting different environments ti dedicated files. Now your webpack config is unsupportable.

**Comment by qdouble** [comment]
@vadimzham I disagree, it's a matter of preference I suppose.

**Comment by vadimzham** [comment]
@qdouble Dedication configs on production/staging/development environments it's not just preference. It's necessity.

**Comment by qdouble** [comment]
@vadimzham that's really just your opinion. There's no major difference between a production and development config that can't be solved with pushing in a few extra plugins based on variables. It's a design choice.

**Comment by sbehun** [comment]
It would be great if we can incorporate tree shaking into the prod AOT build process as well.

https://angular.io/docs/ts/latest/cookbook/aot-compiler.html#!#tree-shaking

**Comment by smith64fx** [comment]
@vadimzham We are already using the webpack starter of @qdouble in production.

Have a look at this sexy & fast website :-) https://secrets.schmid.digital

It uses AoT and is ultra fast. The website is now about 4-5x faster than before! Thanks again to @qdouble . He does awesome work! :)

Check out this webpack starter, if you can't wait longer to use AoT!
https://github.com/qdouble/angular-webpack2-starter

**Comment by vadimzham** [comment]
@smith64fx yea, i have to use AOT on our website too. But looks like i will use angular-cli to do it. Also now i have problems with with several packages of AOT support. So, i have to wait several days(or fix it by myself).

**Comment by smith64fx** [comment]
@vadimzham Yes not every package supports AoT yet.

What i still don't really understand about Angular-cli: Is it a replacement or an extension for webpack? And in the future, what is the way to go?

btw the webpackstarter from @qdouble also uses angular-cli for AoT:

From the package.json file:
```
"compile:aot": "npm run sass && npm run clean:compile && npm run ngc && npm run build:aot",
```

**Comment by vadimzham** [comment]
@smith64fx angular-cli is webpack-based tool that help you to start clear project. It will have much less code than any of `angular2-webpack-starter` or `angular-webpack2-starter` projects. Also it will help you generating new components/modules/etc.

**Comment by itrethan** [comment]
It seems there is no plan to support AOT in this repo based on the previous comments, is that true?

**Comment by katallaxie** [comment]
@itrethan 🎱 No, it is definitely something to support. I hope to do something on that frontier soon.

**Comment by katallaxie** [comment]
@itrethan Have a look at https://github.com/AngularClass/angular2-webpack-starter/tree/aot

**Comment by unsafecode** [comment]
As far as I could do some testing, the best option so far is @ngtools/webpack.

I could make it work without too much effort, and especially without impacting my existing configuration.

**Comment by itrethan** [comment]
@katallaxie Thanks, are you going to merge this with master or just leave the aot part in a separate branch?

**Comment by katallaxie** [comment]
@itrethan we will bring it to master, hopefully with ts and dll.

**Comment by katallaxie** [comment]
@itrethan Have a look at [V6](https://github.com/AngularClass/angular2-webpack-starter/tree/v6) it is a core version of the starter; fully typescript, with aot, and dll support. Though some things to work on still, unit testing, `index.html` production.

**Comment by unsafecode** [comment]
@itrethan I still not get why you reference `ngc` in npm scripts. It is actually ever used? AFAIK, with **@ngtools/webpack** you don't need to leverage ngc itself directly.

Am I missing something?

**Comment by Spittal** [comment]
I was thinking the exact same thing, why is this v6 branch not using `@ngtools/webpack`

**Comment by unsafecode** [comment]
@Spittal ngtools/webpack is already there, you can find it in package.json and webpack.config.ts.

What I was pointing out is whether ngc is still relevant, since the npm script "aot" is not called before "build:aot"

**Comment by katallaxie** [comment]
@unsafecode yep, its there, but not actually used.

**Comment by unsafecode** [comment]
@katallaxie No worries - was just curious to understand whether it was still meaningful, or not.

**Comment by katallaxie** [comment]
@unsafecode I was actually not so sure if it is best to use the plugin. Using `ngc` makes things more transparent and results in less deps?! I would love to have everything as slim as possible. And (vs6)[https://github.com/AngularClass/angular2-webpack-starter/tree/v6] is what I have in mind for my projects. I think that further down the road it would be best to have the starter and some sort of recipes (so, some modules).

**Comment by unsafecode** [comment]
@katallaxie The problem with ngc is it cannot work easily work w/ Sass/Less and pre/post processing steps for HTML templates. AFAIK, it only works if all imported files are already on the filesystem when it runs, which is terribly against what webpack tries to do.

**Comment by katallaxie** [comment]
@unsafecode 🎱 then join the force 🛩 and lets get this running.

**Comment by kjartanvalur** [comment]
any progress @katallaxie @unsafecode  ?

**Comment by unsafecode** [comment]
@kjartanvalur So far, my experimentation has been successful relying on [@ngtools/webpack](https://www.npmjs.com/package/@ngtools/webpack). The repo has been updated with some documentation, but you can find a testing app hidden in their source code here https://github.com/angular/angular-cli/tree/master/tests/e2e/assets/webpack/test-app

**Comment by Spittal** [comment]
Yeah, I'll be honest it seems like we're in a bit of a 'growing pain' area with AOT and Universal. Two things I really want to implement yet there doesn't seem to be a great way to do so without having a HUGE amount of boilerplate. I've been keeping a close eye on this project https://github.com/qdouble/angular-webpack2-starter as it seems very promising.

**Comment by katallaxie** [comment]
@kjartanvalur I'm effectively using [v6](https://github.com/AngularClass/angular2-webpack-starter/tree/v6) for production. I do not know what @gdi2290 plans are for further `dev`.

**Comment by katallaxie** [comment]
@unsafecode we could pull [@ngtools/webpack](https://www.npmjs.com/package/@ngtools/webpack) in v6.

**Comment by Spittal** [comment]
@katallaxie That would be ideal for me, I am really interested in merging v6 into my production apps that use postcss to process styles, which currently isn't support by ngc.

**Comment by gdi2290** [comment]
@katallaxie can you use @ngtools and merge v6

**Comment by qdouble** [comment]
There's still a lot of bugs with @ngtools that's why I'm not using it directly in my repo yet: https://github.com/angular/angular-cli/issues?utf8=%E2%9C%93&q=is%3Aissue%20is%3Aopen%20aot

Using ngc before hand and then compiling it after gets around many of the current bugs

**Comment by katallaxie** [comment]
@gdi2290 @Spittal @qdouble @unsafecode working on `@ngtools/webpack`. I have pushed some things, more work to do. But help is always welcome.

**Comment by qdouble** [comment]
@katallaxie some game breakers for me are
1) ViewEncapsulation.None or Emulated break the build or cause styles not to import properly: 
https://github.com/angular/angular-cli/issues/2584

2) Nested types are often not properly resolved
It's related to this TS bug https://github.com/Microsoft/TypeScript/issues/9944, however if I use my starter where I compile and build as two different processes, the bug doesn't occur

3) It strips custom decorators: https://github.com/angular/angular-cli/issues/2799

Once those 3 issues are solved, I'll switch over to it, but for now it seems like a bigger hassle to use it rather than just using the scripts to process it as separate tasks.

**Comment by katallaxie** [comment]
@qdouble yep, there are some remaining issues. However, can we test `v6` and move it, or stop it?!

**Comment by unsafecode** [comment]
@paralin You're right, this workaround works - but it's actually counterintuitive. I really hope all this bugs are fixed soon, otherwise AoT will remain a mirage for non-trivial projects...

**Comment by katallaxie** [comment]
@unsafecode @paralin AoT uses static code analysis, so you must use `import` statements rather then `require` which are dynamic. There are some other issues with AoT.

**Comment by qdouble** [comment]
@paralin not fully, well that may be a workaround for ViewEncasuplation.None, it's not a good workaround for Emulated. 

Seems it may be a while before it's worth using ngtools when you can just avoid a bunch of the bugs by just compiling with ngc as a separate process.

**Comment by jskrzypek** [comment]
Maybe this is answered somewhere, but I've been wondering about a general strategy like the one suggested here: https://github.com/s-panferov/awesome-typescript-loader/issues/236

Couldn't we just use `"compiler": "@angular/tsc-wrapped"`?

**Comment by paralin** [comment]
@jskrzypek looks promising, try it? :)

**Comment by d3viant0ne** [comment]
Agree with @qdouble here. As it stands right now, I'm loving AoT but I have implemented it professionally in a similar fashion to Anthony, separate tasks executed explicitly. @ngtools/webpack has some outstanding issues that are blocking it's usefulness.

Small implementation details aside, the only really viable way to get AoT working in a meaningful way would be the route @qdouble took. Kind of wish I would have seen his starter prior to doing all that work myself :/ as we ended up in a similar place.

_**If the intent is to get AoT into this starter sooner rather than later,  qdouble/angular-webpack2-starter would be the path forward.**_ 

That said, there are some pretty significant challenges with AoT once you grow your application to anything of production size. AoT is a commitment to building in a pretty specific way, most of which won't be apparent with the small starter application. For instance, anyone using Sass in the normal fashion of using require statements in components for styles are going to have a lot of work to do. 

If we are going to land AoT now, everyone should be ready for the onslaught of questions and issues when people try to pull in AoT enabled #master and their builds explode ( been there ... lol ). 

Also worth noting, if we are going to release AoT support we definitely need to test the AoT builds for PRs or we are going to be breaking master every other day ( been there too ).

**Comment by fxck** [comment]
@katallaxie @d3viant0ne https://github.com/angular/angular-cli/pull/3583

**Comment by itrethan** [comment]
Any updates? This is a must have feature, any serious production app needs to have AOT enabled. People will be move to other starters if this is missing.

**Comment by colinskow** [comment]
I've got it working perfectly. Will submit a PR very soon.

**Comment by shlomiassaf** [comment]
@colinskow @d3viant0ne 
Did you get SCSS to compile with it?

From what I see no /:

Let me know if you don't I have a solution I'm running with.

**Comment by shlomiassaf** [comment]
@colinskow That's not only for SCSS, that's for all resource.

CSS, LESS, SCSS or whatever, they don't go through the loader chain. 

Development output might not be the same as AOT.

I will post a PR soon with suggestion.

**Comment by fxck** [comment]
Is there any issue for that V6 branch? While AOT is a nice step forwards, this starter is still obsolete without DLLs and other stuff.

**Comment by colinskow** [comment]
@fxck #1341

**Comment by d3viant0ne** [comment]
@fxck - Obsolete without a development experience only improvement? How about you open a pull request for whatever `other stuff` is and contribute more than tripe comments? That includes the earlier version of the above comment where you threaten to go somewhere else.

**Comment by fxck** [comment]
@d3viant0ne say what, I'm not sure what the earlier version of that comment was, but I've long since had to move to @qdouble's starter, so I doubt I was threatening **_I_** was gonna go anywhere.

I've been advocating this starter over others for a long time and was closely following the V6 branch(which was blocked by ngtools issues at the time) which I found superior to what I'm using(as it's based on what I'm using). As soon as all the issues I knew about were resolved, I [tried](https://github.com/AngularClass/angular2-webpack-starter/issues/885#issuecomment-267594608) getting a status update on it, didn't get any answer and instead this issue, which was the only [issue V6 was discussed at](https://github.com/AngularClass/angular2-webpack-starter/issues/885#issuecomment-255840442), as far as I know anyway, was closed, which was what prompted my last comment, asking for whether there was, or going to be another issue created for the V6 branch. I didn't get any reply, which is why I wrote [this](https://github.com/AngularClass/angular2-webpack-starter/issues/1330#issuecomment-270340310).

So how about you cut me a slack, I didn't mean anything offensive in a slightest sense, I'm all up for helping finish that branch but without ever getting an answer about your plans going forward, it's kind of hard.

**Comment by shlomiassaf** [comment]
@fxck You are 1 of many, you want DLL, thats cool...
Who said it's good for everyone...

**Comment by paralin** [comment]
Guys calm down, he just was asking where he could find a status update for v6.

People are WAY too quick to get dramatic.

**Comment by colinskow** [comment]
@shlomiassaf cutting development build time in half with no other side effects is good for everyone.

**Comment by fxck** [comment]
@shlomiassaf geez, DLLs are not even the point. The point was that there was a branch of **_THIS_** repo that already has both AOT and DLLs, is utilising @ngtools/webpack, is written a way more modern way, with all the configs being typesafe and written in typescript and all the while @AngularClass [team members](https://github.com/AngularClass/angular2-webpack-starter/issues/1330#issuecomment-270270806) and collaborators are being completely oblivious to the fact. I honestly couldn't care less, as I had to move to qdouble's starter couple of months ago because I simply couldn't proceed without AOT, but it seems like a wasted time, effort and chance to abandon that branch, unless there are reasons for it, which is what I kept asking for.

**Comment by bitbankdev** [comment]
Sorry but does anyone know how to include lodash using @qdouble's repo ? I can not get it run with `import * as _ from 'lodash'` ?

**Comment by fxck** [comment]
`import { where } from 'lodash'` should work just fine, you might need `npm i @types/lodash --save-dev` though

**Comment by dttung-mario** [comment]
@fxck it looks nice but the checker in my terminal keep saying
```
[at-loader] Checking finished with 1 errors
 95% emitting[at-loader] src/app/features/sync/sync.component.ts:2:23
    Cannot find module 'lodash'.
```
and thus I cannot run `yarn run compile`

**Comment by fxck** [comment]
Have you installed `@types/lodash`? Depending on what version of typescript you are running, you might need `types` in tsconfig as well, see https://www.typescriptlang.org/docs/handbook/tsconfig-json.html

**Comment by dttung-mario** [comment]
Oh thank you very much, it works just fine now @fxck !!

**Comment by shlomiassaf** [comment]
@fxck There is a branch with those but the fact that we waited led to better implementation :)

The current setup with AOT is better then `@ngtools`, this is until they catch up then we move back.
The router loader is solid, with AOT and absolute/alias path support.
We should have proper DLL integration without boilerplate in a day or two... 

Patience :)
